docs/specs/2fa-grace-period.md

```diff
@@ -0,0 +1,113 @@
+# 2FA Grace Period
+
+- [Abstract](#abstract)
+- [Configuration](#configuration)
+  - [Global Grace Period](#global-grace-period)
+  - [Per-user Grace Period](#per-user-grace-period)
+  - [Customized Authentication Flow](#customized-authentication-flow)
+
+## Abstract
+
+In order to enforce 2FA, we will provide a grace period for users to enroll in 2FA after they have signed up.
+
+After the grace period, the user will be required to enroll in 2FA before they can sign in.
+
+## Configuration
+
+By default, no grace period is provided. User without 2FA must contact admin to login after [2FA mode](./user-model.md#secondary-authenticator) is set to `required`.
+
+For new users, the grace period starts from the time the user is created, whereas for existing users, the grace period starts from when the grace period is enabled.
+
+### Global Grace Period
+
+Global grace period can be enabled for forcing users to enroll in 2FA upon login instead of blocking them.
+
+```yaml
+authentication:
+  secondary_authentication_mode: "required"
+  secondary_authentication_grace_period:
+    enabled: true
+    end_at: "2021-01-01T00:00:00Z"
+```
+
+### Per-user Grace Period
+
+Regardless of global grace period, specific users can be granted grace period through Admin Portal / GraphQL API.
+
+```gql
+type User {
+  id: ID!
+  # ...
+
+  mfaGracePeriodEndAt: DateTime
+}
+
+type SetMFAGracePeriodInput {
+  userID: ID!
+  endAt: DateTime!
+}
+
+type SetMFAGracePeriodPayload {
+  user: User!
+}
+
+type RemoveMFAGracePeriodInput {
+  userID: ID!
+}
+
+type RemoveMFAGracePeriodPayload {
+  user: User!
+}
+
+type Mutation {
+  setMFAGracePeriod(
+    input: SetMFAGracePeriodInput!
+  ): SetMFAGracePeriodPayload!
+  removeMFAGracePeriod(
+    input: RemoveMFAGracePeriodInput!
+  ): RemoveMFAGracePeriodPayload!
+}
```

Comment:
Follow https://github.com/authgear/authgear-server/pull/4409/files#diff-d2fe537b463363c0561f08dc27a16fd0fcaa38d9462307196bfcd0e32bd33357R93 and document the expected errors in a table.

For this mutation, Invalid MFAGracePeriodInvalid